### ISSUE TITLE
Attention Head Analysis

### ISSUE THREAD

**Issue by jds8** [jds8/SAE4DLM-CE#42]
We are interested in how attention patterns evolve over time. Most layers of a DLM have self-attention sublayers. The attention pattern is the square matrix that indicates how much each token "attends to" the other tokens, i.e. it's a measure of importance. Each row of the matrix sums to 1, so it is a probability distribution. (See graphic below for further explanation) We want to know how the entropy of these distributions changes over diffusion time (temporally) and across layer depth (spatially). The goal for this issue is to generate two plots: average attention entropy versus diffusion time for both masked and unmasked tokens, as shown below. 

We should plot layers at every depth on the same plot, and each such layer should have three lines: all_tokens, masked_tokens, and unmasked_tokens. For the masked plot, the all_tokens line is computed by taking the average entropy of the masked token distributions, i.e. the entropies of those rows corresponding to masked tokens. The unmasked_tokens line is computed by taking only those columns corresponding to unmasked tokens and computing the entropy of the renormalized probability distribution for each row that represents a masked token. The masked_tokens line is the same but takes those columns corresponding to masked tokens. Replicate this for each attention head at every layer depth and plot them all on the same plot. Make each line for a layer the same color, but use different dash marks to distinguish between them. Then do it all again for the unmasked plot.

All of this can be done when the model is invoked on a single text sequence. Ultimately we will want to take these averages across 1000 sequences as well, but let's start with one.

Here is an example of the masked plot (although the colors within a layer should all be the same): 

<img width="567" height="455" alt="Image" src="https://github.com/user-attachments/assets/7c91deb1-31f0-45ce-b394-52a9cdbcdf13" />

These plots were inspired by figure 4 of this paper: https://arxiv.org/pdf/1906.04341.

Graphic:
Each column of the Word Embedding matrix in the following graphic represents a token/word in the sentence "How are you doing?". Because num_tokens = 5, there are five columns. This matrix gets multiplied by Wk, Wq, and Wv, which have the wrong dimensions in the figure, so ignore the number of rows/columns. The product of the Word Embedding matrix, often denoted X, and Wk, generates the Keys matrix (analogously for Wq and the Query matrix). The number of columns in the Keys and Queries matrix equals the number of tokens in the text sequence, again five. These two matrices get multiplied to produce a logits matrix of shape num_tokens x num_tokens, and then a softmax operation is applied over each row to create a set of num_tokens probability distributions, where each row is a probability distribution. In a DLM, some of these rows correspond to masked tokens whereas others correspond to unmasked tokens.
<img width="3399" height="1665" alt="Image" src="https://github.com/user-attachments/assets/b20dfe2d-609d-4955-81dc-d20fab93fccf" />

Useful resources:
1) I didn't really understand attention/transformers until I read this: https://jalammar.github.io/illustrated-transformer/

**Comment by jds8** [comment]
We can start with a single layer:

For a fixed layer ℓ, we want to visualize how the behavior of each attention head changes over diffusion time. At each timestep t, every head h produces an attention matrix where each row corresponds to a token and represents a probability distribution over which other tokens it attends to. For each row, compute its entropy to measure how concentrated or spread out the attention is (low entropy = focused attention, high entropy = diffuse attention). Then average these entropy values across all tokens to get a single number for that head at that timestep.

The resulting plot has diffusion time on the x-axis and average attention entropy on the y-axis, with one line per attention head in the chosen layer. This allows us to see how different heads become more or less focused over time, and whether they exhibit different behaviors during the diffusion process.
